pkg/app/mutate_app.go

```diff
@@ -61,18 +63,92 @@ func (m *Mutator) Mutate(request *v1beta1.AdmissionRequest) ([]mutator.PatchOper
 		return nil, microerror.Maskf(parsingFailedError, "unable to parse app: %#v", err)
 	}
 
+	m.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("mutating app %#q in namespace %#q", appNewCR.Name, appNewCR.Namespace))
+
+	// We check the deletion timestamp because app CRs may be deleted by
+	// deleting the namespace they belong to.
+	if !appNewCR.DeletionTimestamp.IsZero() {
+		m.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("admitted deletion of app %#q in namespace %#q", appNewCR.Name, appNewCR.Namespace))
+		return nil, nil
+	}
+
 	result, err := m.MutateApp(ctx, *appNewCR, *appOldCR)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
+	m.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("applying %d patches to app %#q in namespace %#q", len(result), appNewCR.Name, appNewCR.Namespace))
+
 	return result, nil
 }
 
 func (m *Mutator) MutateApp(ctx context.Context, appNewCR, appOldCR v1alpha1.App) ([]mutator.PatchOperation, error) {
-	return []mutator.PatchOperation{}, nil
+	var result []mutator.PatchOperation
+
+	configPatches, err := m.mutateConfig(ctx, appNewCR, appOldCR)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	if len(configPatches) > 0 {
+		result = append(result, configPatches...)
+	}
+
+	kubeConfigPatches, err := m.mutateKubeConfig(ctx, appNewCR, appOldCR)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	if len(kubeConfigPatches) > 0 {
+		result = append(result, kubeConfigPatches...)
+	}
+
+	return result, nil
 }
 
 func (m *Mutator) Resource() string {
 	return Name
 }
+
+func (m *Mutator) mutateConfig(ctx context.Context, appNewCR, appOldCR v1alpha1.App) ([]mutator.PatchOperation, error) {
```

Comment:
Here too, I can't find `appOldCR` usage.

Reply:
Yes I'll remove it. The admission request has both but we don't need the old CR version.